### ISSUE TITLE
[翻译] The Node.js Event Loop, Timers, and process.nextTick()

### ISSUE THREAD

**Issue by meooxx** [meooxx/blog#20]
[原文:The Node.js Event Loop, Timers, and process.nextTick() ](https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#timers)

个人喜欢中英对照的文章所以， 一定会有中英对照
这篇文章极好的， 闲来无事翻译一波

**Comment by meooxx** [comment]
名次对应关系
     event loop -> 事件循环
     operations -> 操作 （或者一个事件?）
     kernel -> 内核
     callback -> 回调
     queue -> 队列
     phase -> 阶段
  
翻译部分用 *中:* 标记
  
## What is the Event Loop?
The event loop is what allows Node.js to perform non-blocking I/O operations — despite the fact that JavaScript is single-threaded — by offloading operations to the system kernel whenever possible.

Since most modern kernels are multi-threaded, they can handle multiple operations executing in the background. When one of these operations completes, the kernel tells Node.js so that the appropriate callback may be added to the poll queue to eventually be executed. We'll explain this in further detail later in this topic.


>  *中*: event loop 是什么鬼? 

>>event loop 通过尽可能将操作交给内核处理,  来允许nodejs 去执行非阻塞的io 操作——尽管事实上 js是单线程的

>>由于大多数的现代流行的内核都是 多线程的， 所以他们可以在后台执行多个操作。 当其中一个操作完成的时候， 内核就会告诉nodejs, 以便于将适当时机的callback 加入到 poll 队列中，然后最终去执行它. 我们待会将深入讨论上述过程的细节。
 
## Event Loop Explained   
 When Node.js starts, it initializes the event loop, processes the provided input script (or drops into the REPL, which is not covered in this document) which may make async API calls, schedule timers, or call process.nextTick(), then begins processing the event loop.
The following diagram shows a simplified overview of the event loop's order of operations.


 
> *中:* 解释一下event loop

>>当 nodejs 执行的时候, 他将会初始化event loop,  处理那些可能会调用异步API, 调度 timers 或者调用process.nextTick() 的输入脚本(或者放入 REPL，我们这篇文章暂时不涉及REPL), 然后才开始处理 event loop
下面的图解 展示了一个简化的，事件循环中操作顺序的概览

```shell
 # ssd
   ┌───────────────────────────┐
┌─>│           timers          │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
│  │     pending callbacks     │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
│  │       idle, prepare       │
│  └─────────────┬─────────────┘      ┌───────────────┐
│  ┌─────────────┴─────────────┐      │   incoming:   │
│  │           poll            │<─────┤  connections, │
│  └─────────────┬─────────────┘      │   data, etc.  │
│  ┌─────────────┴─────────────┐      └───────────────┘
│  │           check           │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
└──┤      close callbacks      │
   └───────────────────────────┘
note: each box will be referred to as a "phase" of the event loop
注意： 每个box在 eventloop 中被引申为`phase`(译者注释: 阶段)

```

Each phase has a FIFO queue of callbacks to execute. While each phase is special in its own way, generally, when the event loop enters a given phase, it will perform any operations specific to that phase, then execute callbacks in that phase's queue until the queue has been exhausted or the maximum number of callbacks has executed. When the queue has been exhausted or the callback limit is reached, the event loop will move to the next phase, and so on. 

Since any of these operations may schedule more operations and new events processed in the poll phase are queued by the kernel, poll events can be queued while polling events are being processed. As a result, long running callbacks can allow the poll phase to run much longer than a timer's threshold. See the timers and poll sections for more details.

NOTE: There is a slight discrepancy between the Windows and the Unix/Linux implementation, but that's not important for this demonstration. The most important parts are here. There are actually seven or eight steps, but the ones we care about — ones that Node.js actually uses - are those above.

>每一个 `phase` 都有一个 callback 的 FIFO 执行队列(先入先出队列)。 虽然每个 phase  都有其自身的特殊性， 通常, 当event loop 进入到一个 phase 的时候， 他就会执行在那个 phase  中指定的一些操作，然后执行那个 phase 队列中的callback ，直到队列被执行完毕或者达到已执行的 callback 的最大数量值。 当队列被执行完毕或者到达执行callback的限制值的时候，eventloop 将会移动到下一个 phase ，一直这样重复下去

>由于这些操作中可能会调用更多的操作，而且 poll 阶段新的事件也会被内核排列, 依次处理，当有正在处理的 polling(轮询) 事件的时候，当前 poll 事件可以被排列起来。结果是，相较与 timer 的运行时长阈值限制， poll 阶段长时间运行的回调会被允许更长的时间。到 timers 和 poll 的部分查看更多细节

> **注意**： 在window 和 unix/linux 的实现上略有不同, 但是在这个示例中并不重要。最重要部分已在这了。事实上是有7-8个步骤, 但是我们关心的——nodejs 实际使用的那些——全在这上面所述

# Phases Overview
* timers: this phase executes callbacks scheduled by `setTimeout()` and `setInterval()`.
* pending callbacks: executes I/O callbacks deferred to the next loop iteration.
* idle, prepare: only used internally.
* poll: retrieve new I/O events; execute I/O related callbacks (almost all with the exception of close callbacks, the ones scheduled by timers, and `setImmediate()`); node will block here when appropriate.
* check: `setImmediate()` callbacks are invoked here.
* close callbacks: some close callbacks, e.g. `socket.on('close', ...)`.
Between each run of the event loop, Node.js checks if it is waiting for any asynchronous I/O or timers and shuts down cleanly if there are not any.

>phase 概览
>> * timers: 这个阶段执行 setTimeout() 和 setTimeInterval() 调度的回调
>>* pending callbacks: 执行 上次循环延迟到这个阶段执行的 IO 回调
>>* idle, prepare: 只在内部使用.
>>* poll: 获取新的 IO 事件; 执行相关IO的回调(除了 close , timers ，和 `setTimeImmediate()` 等的回调)；node 将会在合适的实际阻塞在这个阶段
>>* check: `setImmediate()` 的回调在这里被调用
>>* close callbacks: 部分的 close 回调，例如。`socket.on('close', ...)`

每次 event loop 运行之间， nodejs 都会检查是否有在等待的异步 IO 或者 timers， 如果没有则彻底关闭

**Comment by meooxx** [comment]
## Phases in Detail
### timers
A timer specifies the threshold after which a provided callback may be executed rather than the exact time a person wants it to be executed. Timers callbacks will run as early as they can be scheduled after the specified amount of time has passed; however, Operating System scheduling or the running of other callbacks may delay them.

Note: Technically, the poll phase controls when timers are executed.

For example, say you schedule a timeout to execute after a 100 ms threshold, then your script starts asynchronously reading a file which takes 95 ms:

>*中:* phase 细节
>> timers
>> 在一个 timer 指定的阈值之后, 其提供的callback 可能被执行， 而不是在一个你期望的确切时间去执行回调。 在过了指定的时间后，可以被调度的时候，timers的回调 将会被尽可能早的执行；但是 操作系统的调度 或者是 其他的callback 都可能让他们被延迟执行

**注意**： 从技术上来说， poll 阶段控制何时 timer 被执行
举个例子说，你调度一个100ms 后执行的 timeout ， 然后你的脚本开始异步的去读取一个需要花费95ms的文件

```js
const fs = require('fs');

function someAsyncOperation(callback) {
  // Assume this takes 95ms to complete
  fs.readFile('/path/to/file', callback);
}

const timeoutScheduled = Date.now();

setTimeout(() => {
  const delay = Date.now() - timeoutScheduled;

  console.log(`${delay}ms have passed since I was scheduled`);
}, 100);


// do someAsyncOperation which takes 95 ms to complete
someAsyncOperation(() => {
  const startCallback = Date.now();

  // do something that will take 10ms...
  while (Date.now() - startCallback < 10) {
    // do nothing
  }
});
```
When the event loop enters the poll phase, it has an empty queue (fs.readFile() has not completed), so it will wait for the number of ms remaining until the soonest timer's threshold is reached. While it is waiting 95 ms pass, fs.readFile() finishes reading the file and its callback which takes 10 ms to complete is added to the poll queue and executed. When the callback finishes, there are no more callbacks in the queue, so the event loop will see that the threshold of the soonest timer has been reached then wrap back to the timers phase to execute the timer's callback. In this example, you will see that the total delay between the timer being scheduled and its callback being executed will be 105ms.

>当event loop到了 poll 阶段，他只有个空的队列(`fs.readFile()` 还没有完成)，所以他将会等待剩下的 ms 数直到最近一个  timer 的阈值到达。 当他等了95ms 后， `fs.readFile()` 完成读取文件， 它的那个花费10ms 才能完成的callback 被加入到 poll 队列然后被执行。 当 这个callback 完成的时候， 队列中没有其他的callback了，所以 event loop 将会查看最近达到阈值的 timer 然后 回到 timers 阶段去执行定时器的回调。在这个例子中，你将看到在调度  timer 开始到它的 callback 被执行之间总的延迟将会是 105ms。

Note: To prevent the poll phase from starving the event loop, libuv (the C library that implements the Node.js event loop and all of the asynchronous behaviors of the platform) also has a hard maximum (system dependent) before it stops polling for more events.

>*注意*:  为了防止poll 阶段一直阻塞在空的eventloop中， 在libuv 停止轮询更多事件前，  libuv( 实施 nodejs event loop 以及平台的所有异步行为的C库) 也有一个 硬性最大值(依赖于系统的)。

## pending callbacks
This phase executes callbacks for some system operations such as types of TCP errors. For example if a TCP socket receives ECONNREFUSED when attempting to connect, some *nix systems want to wait to report the error. This will be queued to execute in the pending callbacks phase.

>pending callbacks
>>这个阶段是为了执行一些系统操作回调，像一些 TCP errors 的类型啊。例如 当Tcp尝试连接的时候， 如果一个TCP socket(套接字) 收到一个 `ENCONNERREFUSED`  一些 *nix 系统想等待去报告这个错误。 这将被有序的放在 pending callback 阶段执行

## poll  
The poll phase has two main functions:

* Calculating how long it should block and poll for I/O, then
* Processing events in the poll queue.
> poll
>> poll 阶段有俩个主要功能

>>* 计算应该为 I/O 阻塞 和 轮询 的时长,  然后
>>* 处理 poll 阶段的事件

When the event loop enters the poll phase and there are no timers scheduled, one of two things will happen:

* If the poll queue is not empty, the event loop will iterate through its queue of callbacks executing them synchronously until either the queue has been exhausted, or the system-dependent hard limit is reached.

* If the poll queue is empty, one of two more things will happen:
    * If scripts have been scheduled by setImmediate(), the event loop will end the poll phase and continue to the check phase to execute those scheduled scripts.

     * If scripts have not been scheduled by setImmediate(), the event loop will wait for callbacks to be added to the queue, then execute them immediately.

>*中文：* 当 eventloop 进入到 poll 阶段到时候且没有已调度的 timer , 会发生下面两种情况之一:
>> * 如果 队列不是空的，eventloop 将会遍历他的回调队列同步的执行他们，直到队列被执行完毕或者达到基于系统相关的硬性限制
>>* 如果poll 队列是空的， 将要发生另外的俩件事之一:
>>    * 如果有被 `setImmediate()` 调度的脚本， eventloop 将会结束 poll 阶段，继续到 check 阶段去执行那些被调度的脚本
>>     * 如果没有被 `setImmediate()`调度的脚本, eventloop 将会等待一些callback 被加入到队列， 然后立即执行他们

Once the poll queue is empty the event loop will check for timers whose time thresholds have been reached. If one or more timers are ready, the event loop will wrap back to the timers phase to execute those timers' callbacks.

> *中：*  一旦poll队列是空的， eventloop 将会检查已经到达阈值的一些 timer. 如果一个或多个 timer 已经准备好，eventloop 将会转到 timers 阶段去执行这些 timers的回调

**Comment by meooxx** [comment]
## check
This phase allows a person to execute callbacks immediately after the poll phase has completed. If the poll phase becomes idle and scripts have been queued with setImmediate(), the event loop may continue to the check phase rather than waiting.

setImmediate() is actually a special timer that runs in a separate phase of the event loop. It uses a libuv API that schedules callbacks to execute after the poll phase has completed.

Generally, as the code is executed, the event loop will eventually hit the poll phase where it will wait for an incoming connection, request, etc. However, if a callback has been scheduled with setImmediate() and the poll phase becomes idle, it will end and continue to the check phase rather than waiting for poll events.
>*中:*  ## check
>>在poll 阶段完成后, 在这个check阶段允许人为去立即执行一些回调。如果 poll 阶段变空闲了而且有使用`setImmediate()` 被放置队列中的脚本， event loop 可能继续在 check 阶段而不是 一直等待

>>`setImmediate()` 是运行在 event loop 的一个独立阶段的 特殊timer。在poll 阶段已完成后, 他使用 libuv API  调度callback 去执行

>> 一般的，当 代码被执行的时候， event loop 将最终会到达 等待传入连接， request ...的 poll 阶段。但如果 有一个已被`setImmediate()`调度的callback 且 poll 阶段已经变得空闲，他将会结束并进入到check 阶段而不是一直等待 轮询事件

**Comment by meooxx** [comment]
## close callbacks
If a socket or handle is closed abruptly (e.g. socket.destroy()), the 'close' event will be emitted in this phase. Otherwise it will be emitted via process.nextTick().

setImmediate() vs setTimeout()
setImmediate and setTimeout() are similar, but behave in different ways depending on when they are called.

* setImmediate() is designed to execute a script once the current poll phase completes.
* setTimeout() schedules a script to be run after a minimum threshold in ms has elapsed.
>*中文：* ## close callback
> 如果一个 socket 或者 一个 handle 被突然的关闭(例如, socket.destory()), 这个 'close' 事件会在这个阶段被触发。否则它会被process.nextTick() 触发

>setImmediate() 和 setTimeout()
>setImmediate 和 setTimeout() 是相似的, 它们的 行为在某些方面是不同的， 这取决于他们被调用的时机。
> * setImmediate() 被设计成，当前 poll 阶段完成就立刻执行一段脚本。
>* setTimeout()  调度的脚本， 会在一个最低 ms 阈值过去之后运行

The order in which the timers are executed will vary depending on the context in which they are called. If both are called from within the main module, then timing will be bound by the performance of the process (which can be impacted by other applications running on the machine).

For example, if we run the following script which is not within an I/O cycle (i.e. the main module), the order in which the two timers are executed is non-deterministic, as it is bound by the performance of the process:

> timers 执行的顺序是变化的，取决与它们被调用的上下文中。如果都在主模块中被调用，那么时间将会被进程(可能被机器上的其他应用影响)的性能限制  

>例如，如果我们运行下面的不在 I/O 周期内(即主模块中)的脚本， 这俩个timers 的执行顺序是不确定的，因为它受到进程性能的限制

```javascript
// timeout_vs_immediate.js
setTimeout(() => {
  console.log('timeout');
}, 0);

setImmediate(() => {
  console.log('immediate');
});
```
```shell
$ node timeout_vs_immediate.js
timeout
immediate

$ node timeout_vs_immediate.js
immediate
timeout

```

However, if you move the two calls within an I/O cycle, the immediate callback is always executed first:
>但是，如果你把这俩个调用移动到 I/O 周期中，immediate 的回调 总是先执行

```js
// timeout_vs_immediate.js
const fs = require('fs');

fs.readFile(__filename, () => {
  setTimeout(() => {
    console.log('timeout');
  }, 0);
  setImmediate(() => {
    console.log('immediate');
  });
});
```
``` 
$ node timeout_vs_immediate.js
immediate
timeout

$ node timeout_vs_immediate.js
immediate
timeout
```
The main advantage to using setImmediate() over setTimeout() is setImmediate() will always be executed before any timers if scheduled within an I/O cycle, independently of how many timers are present.
>使用 `setImmediate() `胜过 `setTimeout()`的主要优势是，如果是被调度在 I/O周期中，则   `setImmediate() `执行在 任何timer之前 ， 不管有多少的 timers 存在

**Comment by meooxx** [comment]
# process.nextTick()
## Understanding process.nextTick()
You may have noticed that process.nextTick() was not displayed in the diagram, even though it's a part of the asynchronous API. This is because process.nextTick() is not technically part of the event loop. Instead, the nextTickQueue will be processed after the current operation completes, regardless of the current phase of the event loop.

Looking back at our diagram, any time you call process.nextTick() in a given phase, all callbacks passed to process.nextTick() will be resolved before the event loop continues. This can create some bad situations because it allows you to "starve" your I/O by making recursive process.nextTick() calls, which prevents the event loop from reaching the poll phase.

># *中文: *process.nextTick()
>## 理解 process.nextTick()

>你可能已经注意到 process.nextTick() 不在图表中的，即使它是异步API 一部分。这是因为 process.nextTick()  技术上不是 event loop 的一部分。 相反，nextTickQueue 将在当前操作完成之后才被处理，不管 event loop 的当前阶段

>回过头看我们的图解, 你在给出的阶段任何时机调用process.nextTick()，在event loop 继续之前，所有传递给 process.nextTick()的回调将会被 resolve(译者注: 执行/解决/解析- 这个resovle 不太好翻译)。 这可能造成一个不好的状况,  因为它通过递归调用 process.nextTick(),  它会让你 "starve" (暂译: 阻塞卡死)你的 I/O ，这将阻止 event loop 到达 poll 阶段


Why would that be allowed?
Why would something like this be included in Node.js? Part of it is a design philosophy where an API should always be asynchronous even where it doesn't have to be. Take this code snippet for example:

```js
function apiCall(arg, callback) {
  if (typeof arg !== 'string')
    return process.nextTick(callback,
                            new TypeError('argument should be string'));
}
```

> 为什么会被允许这么干呢？
>为什么上述情况会被包括进 nodejs?  这一部分是一个设计理念，在于 一个API 应该始终是异步的虽然他不必须是。以下面代码为例子:

The snippet does an argument check and if it's not correct, it will pass the error to the callback. The API updated fairly recently to allow/ passing arguments to process.nextTick() /allowing it to take any arguments passed after the callback /to be propagated as the arguments to the callback/ so you don't have to nest functions.

>这段代码对一个 argument 作检查如果它不正确的话，它会把错误传递给 callback。这个 API 最近的更新中允许传一些参数到 process.nextTick() ，这允许接受任何被传递到 callback 位置之后的参数 ，将它作为参数传递给callback，所以你不必去嵌套functions了。

What we're doing is passing an error back to the user but only after we have allowed the rest of the user's code to execute. By using process.nextTick() we guarantee that apiCall() always runs its callback after the rest of the user's code and before the event loop is allowed to proceed. To achieve this, the JS call stack is allowed to unwind then immediately execute the provided callback which allows a person to make recursive calls to process.nextTick() without reaching a RangeError: Maximum call stack size exceeded from v8.

>我们正在做的是传回 error 给用户，但是这只在我们已经让用户代码的其余部分执行完之后。通过使用 process.nextTick() 我们保证了 apiCall() 总是在用户代码其余部分之后和 event loop 继续之前，运行它的回调函数。 为了实现这个，JS 调用栈展开(stack unwind)然后立即执行提供的回调函数，这允许一个人去递归的调用 process.nextTick() 而不会触发(达到)`RangeError: Maximum call stack size exceeded from v8`(译者注: RangeError: v8的最大调用栈大小超出)。

This philosophy can lead to some potentially problematic situations. Take this snippet for example:
>这个理念可能导致一些潜在的问题状况。拿这个片段举例:

```js
let bar;

// this has an asynchronous signature, but calls callback synchronously
// 译者注: https://developer.mozilla.org/en-US/docs/Glossary/Signature/Function
// 签名: 一个函数输入输出及其类型，抛出或者传回的异常，面向对象的可用性信息(public, prototype..)
// 这有个异步的签名(参数)，却是同步的调用回调
function someAsyncApiCall(callback) { callback(); }

// the callback is called before `someAsyncApiCall` completes.
// 回调被调用在 `someAsyncApiCall` 完成之前
someAsyncApiCall(() => {
  // since someAsyncApiCall has completed, bar hasn't been assigned any value
  console.log('bar', bar); // undefined
});

bar = 1;
```

The user defines someAsyncApiCall() to have an asynchronous signature, but it actually operates synchronously. When it is called, the callback provided to someAsyncApiCall() is called in the same phase of the event loop because someAsyncApiCall() doesn't actually do anything asynchronously. As a result, the callback tries to reference bar even though it may not have that variable in scope yet, because the script has not been able to run to completion.

> 用户定义 someAsyncApiCall() 让它有一个异步的签名，但是它实际上是同步的操作。当它被调用时， 提供给 someAsyncApiCall() 的回调是被调用在同一个event loop 阶段，这是因为 someAsyncApiCall() 实际上没有异步的去做任何事情。结果就是，callback 尝试去引用 bar ，即使它在作用域中可能还没有这个变量,，因为脚本还没能运行完成。
By placing the callback in a process.nextTick(), the script still has the ability to run to completion, allowing all the variables, functions, etc., to be initialized prior to the callback being called. It also has the advantage of not allowing the event loop to continue. It may be useful for the user to be alerted to an error before the event loop is allowed to continue. Here is the previous example using process.nextTick():
> 通过把callback 放到 process.nextTick()，脚本依然可以运行完成，允许所有的变量，functions 等等， 被初始化在callback 调用之前。 它也具有不让 event loop 继续的优势。这可能对于在让 event loop 继续之前去警告错误给用户是有用的。这有个先前使用 process.nextTick() 例子
```js
let bar;

function someAsyncApiCall(callback) {
  process.nextTick(callback);
}

someAsyncApiCall(() => {
  console.log('bar', bar); // 1
});

bar = 1;
```

Here's another real world example:
>这是另外实际中的例子

```js
const server = net.createServer(() => {}).listen(8080);

server.on('listening', () => {});
```
When only a port is passed, the port is bound immediately. So, the 'listening' callback could be called immediately. The problem is that the .on('listening') callback will not have been set by that time.

To get around this, the 'listening' event is queued in a nextTick() to allow the script to run to completion. This allows the user to set any event handlers they want.

> 当只有传递端口时，会立即绑定端口，所以，'listening'  的 callback 可能被立即调用。 问题是 .on('listening') 那时候还没被设置好呢

>为了绕过这个情况， 'listening' 事件会被排在一个 `nextTick()` 中以让脚本运行完成。这允许用户去设置任何他们想要的事件处理器

**Comment by meooxx** [comment]
# `process.nextTick()` vs `setImmediate()`
We have two calls that are similar as far as users are concerned, but their names are confusing.

* process.nextTick() fires immediately on the same phase
* setImmediate() fires on the following iteration or 'tick' of the event loop

>就用户而言，我们有俩个相似的调用，但是他们的名字让人迷惑。
>* process.nextTick()  立即触发在相同阶段
>* setImmediate() 触发在接下来的 event loop 的迭代或者 'tick' 中

In essence, the names should be swapped. process.nextTick() fires more immediately than setImmediate(), but this is an artifact of the past which is unlikely to change. Making this switch would break a large percentage of the packages on npm. Every day more new modules are being added, which means every day we wait, more potential breakages occur. While they are confusing, the names themselves won't change.

> 实质上，他们名字应该被交换 process.nextTick() 比 setTimmidiate() 触发的更加及时，但这是个过去人为产物(译者注: 理解成历史遗留原因)，不太可能去改变。进行此交换的话，将会导致很大比例的 npm 包挂掉。每一天都有更多的包的加入，就意味着我们要多一天的等待，则更多的破坏会出现。 尽管我们感到疑惑，但它们的名字就是不会更改。

*We recommend developers use setImmediate() in all cases because it's easier to reason about (and it leads to code that's compatible with a wider variety of environments, like browser JS.)*
>我们推荐开发者在所有场景中使用 `setImmediate()` ，是因为它更容易理解(而且可以让代码更好的兼容其他的环境，例如浏览器 js)

# Why use process.nextTick()?
There are two main reasons:

* Allow users to handle errors, cleanup any then unneeded resources, or perhaps try the request again before the event loop continues.

* At times it's necessary to allow a callback to run after the call stack has unwound but before the event loop continues.
># 为什么使用  process.nextTick() ?
> 有俩个主要原因
>* 让用户去处理错误，清除干净任何不再需要的资源，或者在 event loop 继续之前，再去尝试请求
>* 有时需要运行回调在调用栈展开后和 eventloop 继续之前

One example is to match the user's expectations. Simple example:
>一个符合用户预期例子
```js
const server = net.createServer();
server.on('connection', (conn) => { });

server.listen(8080);
server.on('listening', () => { });
```
Say that listen() is run at the beginning of the event loop, but the listening callback is placed in a setImmediate(). Unless a hostname is passed, binding to the port will happen immediately. For the event loop to proceed, it must hit the poll phase, which means there is a non-zero chance /that a connection could have been received allowing the connection event to be fired before the listening event.

>假设 listen() 运行在 event loop 的开始，但是 'listening' 的 callback 被放置在 setImmediate() 中。除非hostname 被传递，绑定到端口才会立即发生。要 event loop 继续进行，就必须到达 poll 阶段。这意味在 'listening' 事件之前，有可能接收到一个 connection 事件， 而且还会让 connection 的事件触发     

Another example is running a function constructor that was to, say, inherit from EventEmitter and it wanted to call an event within the constructor:
> 另外一个例子是运行 一个构造函数也就是说，继承 EventEmitter 类而且它想在构造函数中调用一个 event：
```js
const EventEmitter = require('events');
const util = require('util');

function MyEmitter() {
  EventEmitter.call(this);
  this.emit('event');
}
util.inherits(MyEmitter, EventEmitter);

const myEmitter = new MyEmitter();
myEmitter.on('event', () => {
  console.log('an event occurred!');
});
```
You can't emit an event from the constructor immediately/because the script will not have processed to the point /where the user assigns a callback to that event. So, within the constructor itself, you can use process.nextTick() to set a callback to emit the event after the constructor has finished, which provides the expected results:
>你不能从一个构造函数立即的触发一个事件，因为脚本没处理(运行)到 用户分配 callback 给 event 的那个地方。所以在构造函数本身(译者注: 或者内部)，你可以使用 process.nextTick() 去设置一个 callback， 在构造函数完成之后去触发 event，这提供了一个预期结果。 

```js
function MyEmitter() {
  EventEmitter.call(this);

  // use nextTick to emit the event once a handler is assigned
  process.nextTick(() => {
    this.emit('event');
  });
}
util.inherits(MyEmitter, EventEmitter);

const myEmitter = new MyEmitter();
myEmitter.on('event', () => {
  console.log('an event occurred!');
});
```
